domain/relation/service/relation.go

```diff
@@ -959,3 +980,79 @@ func (s *Service) SetRelationUnitSettings(
 	}
 	return s.st.SetRelationUnitSettings(ctx, relationUnitUUID, settings)
 }
+
+// ImportRelations sets relations imported in migration.
+func (s *Service) ImportRelations(ctx context.Context, args relation.ImportRelationsArgs) error {
+	for _, arg := range args {
+		relUUID, err := s.importRelation(ctx, arg)
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		for _, ep := range arg.Endpoints {
+			err = s.importRelationEndpoint(ctx, relUUID, ep)
+			if err != nil {
+				return errors.Capture(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Service) importRelation(ctx context.Context, arg relation.ImportRelationArg) (corerelation.UUID, error) {
+	var relUUID corerelation.UUID
+
+	eps := arg.Key.EndpointIdentifiers()
+	var err error
+
+	switch len(eps) {
+	case 1:
+		// Peer relations are implicitly imported during migration of applications
+		// during the call to CreateApplication.
+		relUUID, err = s.st.GetPeerRelationUUIDByEndpointIdentifiers(ctx, eps[0])
+		if err != nil {
+			return relUUID, errors.Errorf("getting peer relation %d by endpoint %q: %w", arg.ID, eps[0], err)
+		}
+	case 2:
+		relUUID, err = s.st.SetRelationWithID(ctx, eps[0], eps[1], uint64(arg.ID))
+		if err != nil {
+			return relUUID, errors.Capture(err)
+		}
+	default:
+		return relUUID, errors.Errorf("unexpected number of endpoints %d for %q", len(eps), arg.Key)
+	}
+	return relUUID, nil
+}
+
+func (s *Service) importRelationEndpoint(ctx context.Context, relUUID corerelation.UUID, ep relation.ImportEndpoint) error {
+	appID, err := s.st.GetApplicationIDByName(ctx, ep.ApplicationName)
+	if err != nil {
+		return err
+	}
+
+	err = s.st.SetRelationApplicationSettings(ctx, relUUID, appID, settingsMap(ep.ApplicationSettings))
+	if err != nil {
+		return err
+	}
+	for unitName, settings := range ep.UnitSettings {
+		err = s.st.EnterScope(ctx, relUUID, unit.Name(unitName), settingsMap(settings))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func settingsMap(in map[string]interface{}) map[string]string {
+	return transform.Map(in, func(k string, v interface{}) (string, string) {
+		return k, fmt.Sprintf("%v", v)
```

Comment:
Simons comment from my PR on this still applys here https://github.com/juju/juju/pull/19602#discussion_r2058916652:
He was saying it should be this instead of `fmt.Sprintf`:
```go
switch v.(type) {
case string:
  return v
default:
  return errors.Errorf("...")
}
```